donations/management/commands/check_missing_donations.py

```diff
@@ -0,0 +1,94 @@
+#
+# Freesound is (c) MUSIC TECHNOLOGY GROUP, UNIVERSITAT POMPEU FABRA
+#
+# Freesound is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# Freesound is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#     See AUTHORS file.
+#
+
+import datetime
+import logging
+import requests
+import json
+from urlparse import parse_qs
+from django.conf import settings
+from donations.models import Donation, DonationCampaign
+from django.core.management.base import BaseCommand
+
+logger = logging.getLogger("web")
+
+
+class Command(BaseCommand):
+    help = 'Synchronize Paypal donations'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-d', '--days',
+            action='store_true',
+            dest='days',
+            default=1,
+            help='Use this option to get the donations older than 1 day.')
+
+    def handle(self, **options):
+        days = options['days']
+        logger.info("Synchronizing donations from paypal")
+
+        td = datetime.timedelta(days=days)
+        start = datetime.datetime.now() - td
+
+        one_day = datetime.timedelta(days=1)
+        stop_date = datetime.datetime.now() - one_day
+        params = {
+            'METHOD': 'TransactionSearch',
+            'STARTDATE': start,
+            'ENDDATE': start + one_day,
+            'VERSION': 94,
+            'USER' : settings.PAYPAL_USERNAME,
+            'PWD': settings.PAYPAL_PASSWORD,
+            'SIGNATURE': settings.PAYPAL_SIGNATURE
+        }
+        while stop_date >= start:
+            req = requests.post(settings.PAYPAL_PAYMENTS_API_URL, data=params)
+            raw_rsp = parse_qs(req.text)
+            if raw_rsp['ACK'] == ['Success']:
+                campaign = DonationCampaign.objects.order_by('date_start').last()
+                Donation._meta.get_field('created').auto_now_add = False
+                for i in range(100):
```

Comment:
Why this 100 here?

Reply:
Because they return maximum 100 items. And I think we are not going to have more than 100 donations per day :)

Reply:
@andrebola don't they return the actual number of results? Then you don't have to do this ugly hack of

`if ('L_EMAIL%d' % i) in raw_rsp:`

Reply:
The response has the form: 
`L_TRANSACTIONID0=9XJ88717E86951234&L_TRANSACTIONID1=34A53358N84361234` . You can check an example here:
https://developer.paypal.com/docs/classic/express-checkout/ht_searchRetrieveTransactionData-curl-etc/

Maybe someone has a better way of doing it?

Reply:
Dear god have mercy that's messy. No support for JSON or something more.... sane?! I guess not... I mean I would say that the more sane way is perhaps to filter out the right keys first and then iterate over those? Something like.....

```python
filtered_keys = (key for raw_rsp.keys() if key.starts_with("L_EMAIL"))
for email addres in filtered_keys:
  ...
```

perhaos?

BTW, the docs say: "This integration method is Deprecated as of January 1, 2017"

Reply:
( https://developer.paypal.com/docs/api/payments/#payment_list perhaps? )

Reply:
yeah, but I couldn't find a way of doing it with the REST API. So I guess they will maintain it for a while. 
We should remove Paypal and leave only Stripe!

Reply:
the payments API returns no results, I think it only works if the payment was created using the API.

Reply:
so what's the conclusion about this? we either use a deprecated method or have no other option?

Reply:
Can you clarify on this? Don't we have a stable way to get all the payments? Also regarding the maximum of 100, what happens if there are more? I think it would still be nice to retroactively create donation objects since we deployed new donations things. Therefore the first run of this command would be with a high `--days` number. There will certainly be more that 100. What will happen? Also for this first run I'd disable logging to graylog as messages will be misplaced. Once we run this command daily it's not too bad if we have some donation messages that appear in a different hour, but for the first run we would be adding a lot of donations at the time the command is run. Maybe we could add an option to not send to graylog?

Reply:
@andrebola just pinging you because this comment is buried in the thread and I'm not sure whether giuthub will notify :)